testsuite/MDAnalysisTests/analysis/test_lineardensity.py

```diff
@@ -43,138 +46,219 @@ def test_invalid_grouping():
+    expected_segments = np.array([sum(eval(f'seg.atoms.{prop}')) for seg in u.segments])
+
+    return expected_atoms, expected_residues, expected_segments
+
+def find_Centres(groups): # Always based on Centre of Mass (NOT Centre of Charge)
+    centres = []
+    for group in groups:
+        total_Mass = sum(group.atoms.masses)
+        if total_Mass != 0:
+            centres.append(np.sum(group.atoms.positions.transpose() * abs(group.atoms.masses), axis = 1) / total_Mass)
+        elif total_Mass == 0:
+            centres.append(np.sum(group.atoms.positions.transpose() * abs(group.atoms.masses), axis = 1) / len(group.atoms)) ############## NOT NEDEDEDDED 
+
+    return np.array(centres)
+
+def calc_Densities(u, prop = 'masses', spliceLen = 0.25): # Property can be 'masses' or 'charges'
+
+    propShort = 'mass'
+    if prop == 'charges': propShort = 'charge'
+
+    ### Atoms
+    expected_atoms = np.zeros((3, int(u.dimensions[0] // spliceLen))).astype(float) # Works for cubic Universe
+    for atom in u.atoms:
+        for i in range(3):
+            expected_atoms[i][int(atom.position[i] // spliceLen)] += eval(f'atom.{propShort}')
+
+
+
+    _,residue_Totals,segment_Totals = calc_Prop(u, prop) # Total of Charge OR Mass
+    ### Residues
+    expected_residues = np.zeros((3, int(u.dimensions[0] // spliceLen))).astype(float)
+    residue_Centres = find_Centres(u.residues)
+
+
+    for i in range(len(residue_Centres)):
+        for j in range(3):
+            expected_residues[j][int(residue_Centres[i][j] // spliceLen)] += residue_Totals[i]
+
+    ### Segments
+    expected_segments = np.zeros((3, int(u.dimensions[0] // spliceLen))).astype(float)
+    segment_Centres = find_Centres(u.segments)
+
+
+    for i in range(len(segment_Centres)):
+        for j in range(3):
+            expected_segments[j][int(segment_Centres[i][j] // spliceLen)] += segment_Totals[i]
+    
+    
+    # Scaling based on splice volumes & converting units
+    for i in range(3):
+        expected_atoms[i,:] /= spliceLen * u.dimensions[(i + 1) % 3] * u.dimensions[(i + 2) % 3]
+        expected_residues[i,:] /= spliceLen * u.dimensions[(i + 1) % 3] * u.dimensions[(i + 2) % 3]
+        expected_segments[i,:] /= spliceLen * u.dimensions[(i + 1) % 3] * u.dimensions[(i + 2) % 3]
+    expected_atoms /= constants['N_Avogadro'] * 1e-24 # To be consistent with lineardensity.py
+    expected_residues /= constants['N_Avogadro'] * 1e-24 # To be consistent with lineardensity.py
+    expected_segments /= constants['N_Avogadro'] * 1e-24 # To be consistent with lineardensity.py
+
+    return expected_atoms, expected_residues, expected_segments
+
+### Creating the Expected Values ###
+
+spliceLen = 0.25
+
+universes = []
+groupings = ['atoms', 'residues', 'segments']
+# Will be [[Atoms, Residues, Segments (of Universe 1)], [... of Universe 2], ...]
+expected_masses = [] 
+expected_charges = []
+expected_xmass = []
+expected_xcharge = []
+
+for system in test_Systems:
```

Comment:
With functions in L51, you can remove the `eval()` and run system() in L207. I would also discourage using such a generic name `system` for a variable, esp. since it's callable. Could mistaken with something more common like `os.system`.